rotkehlchen/accounting/cost_basis/base.py

```diff
@@ -98,17 +100,77 @@ def __str__(self) -> str:
+        We can't return here Tuple of AssetAcquisitionEvents as we need to return
+        the first event each time but _acquisitions may be not modified between iterations.
+        """
+        while len(self._acquisitions) > 0:
+            yield self._acquisitions[0]
+
+    def get_acquisitions(self) -> Tuple[AssetAcquisitionEvent, ...]:
+        """Returns read-only _acquisitions"""
+        return tuple(self._acquisitions)
+
+    def consume_result(self, used_amount: FVal) -> None:
+        """This function should be used to consume results of the
+        currently processed event (received from __next__)
+        The current event's remaining_amount will be decreased by used_amount
+        If event's remaining_amount will become ZERO, the event will be deleted
+        May raise:
+        - IndexError if the method was called when acquisitions were empty
+        """
+        # this is a temporary assertion to test that new accounting tools work properly.
+        # Written on 06.06.2022 and can be removed after a couple of months if everything goes well
+        assert ZERO <= used_amount <= self._acquisitions[0].remaining_amount, \
+            f'Used amount must be in the interval [0, {self._acquisitions[0].remaining_amount}] but it was {used_amount}'  # noqa: E501
+
+        self._acquisitions[0].remaining_amount -= used_amount
+        if self._acquisitions[0].remaining_amount == ZERO:
+            self._acquisitions.popleft()
+
+
+class FIFOAcquisitionsOrder(BaseAcquisitionsOrder):
+    """Accounting in FIFO (first-in-first-out) order"""
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        self._acquisitions.append(acquisition)
+
+
+class LIFOAcquisitionsOrder(BaseAcquisitionsOrder):
+    """Accounting in LIFO (last-in-first-out) order"""
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        self._acquisitions.appendleft(acquisition)
+
+
 class CostBasisEvents:
-    used_acquisitions: List[AssetAcquisitionEvent] = field(init=False)
-    acquisitions: List[AssetAcquisitionEvent] = field(init=False)
-    spends: List[AssetSpendEvent] = field(init=False)
+    used_acquisitions: List[AssetAcquisitionEvent]
+    acquisitions_manager: BaseAcquisitionsOrder
+    spends: List[AssetSpendEvent]
 
-    def __post_init__(self) -> None:
-        """Using this since can't use mutable default arguments"""
-        self.used_acquisitions = []
-        self.acquisitions = []
+    def __init__(self, cost_basis_method: CostBasisMethod) -> None:
+        """This class contains data about acquisitions and spends. `acquisitions` field contains
+        custom Iterable that provides acquisitions in the order defined by `cost_basis_method`
+        """
+        if cost_basis_method == CostBasisMethod.FIFO:
+            self.acquisitions_manager = FIFOAcquisitionsOrder()
+        elif cost_basis_method == CostBasisMethod.LIFO:
+            self.acquisitions_manager = LIFOAcquisitionsOrder()
```

Comment:
No is because this case won't hit in production since it is required to modify this if when you add a new accounting order option (and it simply won't work without modifying it)